pyomo/contrib/cp/interval_var.py

```diff
@@ -0,0 +1,202 @@
+
+
+class IntervalVarLength(ScalarVar):
+    """This class defines the abstract interface for a single variable
+    denoting the length of an IntervalVar"""
+
+    __slots__ = ()
+
+    def __init__(self):
+        super().__init__(domain=Integers, ctype=IntervalVarLength)
+
+    def get_associated_interval_var(self):
+        return self.parent_block()
+
+
+class IntervalVarPresence(ScalarBooleanVar):
+    """This class defines the abstract interface for a single Boolean variable
+    denoting whether or not an IntervalVar is scheduled"""
+
+    __slots__ = ()
+
+    def __init__(self):
+        super().__init__(ctype = IntervalVarPresence)
+
+    def get_associated_interval_var(self):
+        return self.parent_block()
+
+
+class IntervalVarData(_BlockData):
+    """This class defines the abstract interface for a single interval variable.
+    """
+    # We will put our four variables on this, and everything else is off limits.
+    _Block_reserved_words = Any
+
+    def __init__(self, component=None):
+        _BlockData.__init__(self, component)
+
+        with self._declare_reserved_components():
+            self.is_present = IntervalVarPresence()
+            self.start_time = IntervalVarStartTime()
+            self.end_time = IntervalVarEndTime()
+            self.length = IntervalVarLength()
+
+    @property
+    def optional(self):
+        # We only store this information in one place, but it's kind of annoying
+        # to have to check if the BooleanVar is fixed, so this way you can ask
+        # the IntervalVar directly.
+        return not self.is_present.fixed
+
+    @optional.setter
+    def optional(self, val):
+        if type(val) is not bool:
+            raise ValueError(
+                "Cannot set 'optional' to %s: Must be True or False." % val)
+        if val:
+            self.is_present.unfix()
+        else:
+            self.is_present.fix(True)
+
+
+@ModelComponentFactory.register("Interval variables for scheduling.")
+class IntervalVar(Block):
+    """An interval variable, which may be defined over an index.
+
+    Args:
+        start (tuple of two integers): Feasible range for the
+            interval variable's start time
+        end (tuple of two integers, optional): Feasible range for the
+            interval variables end time
+        length (integer or tuple of two integers, optional): Feasible
+            range for the length of the interval variable
+        optional (boolean, optional) : If False, the interval variable must
+            be scheduled. Otherwise the interval variable is optionally
+            present. Default behavior is 'False'
+        name (str, optional): Name for this component.
+        doc (str, optional): Text describing this component.
+    """
+
+    _ComponentDataClass = IntervalVarData
+
+    def __new__(cls, *args, **kwds):
+        if cls != IntervalVar:
+            return super(IntervalVar, cls).__new__(cls)
+        if args == ():
```

Comment:
This should check `if not args or (args[0] is UnindexedComponent_set and len(args)==1):`